### PR TITLE
Begin React documentation

### PR DESCRIPTION
**Pull request by BenAtEventbrite**
Starting off the React documentation with:

- Component structure
- JSX styling

Pretty much all of the rules or eslint-enforced.

### CODE REVIEW

File: react/README.md [hunk 1]
```diff
@@ -18,3 +29,259 @@ For more info, see [Getting started with React](https://facebook.github.io/react
...
+
+_NOTE:_ There is a common practice to use stateless/pure functions over ES6 classes for components without any internal state (eslint: [`react/prefer-stateless-function`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md)), but we are choosing to stick with ES6 classes for a few reasons:
+
+- When the component does more than just render `props`, such as attaching callback handlers, the stateless function can become unnecessarily complex with nested functions
+- [`propTypes`](#component-propTypes) are defined _within_ the ES6 class, but have to be defined as additional properties _outside_ of a stateless function
+- Using ES6 classes for the main/default component help differentiate it from [helper components](#)
+
+## Component organization
+
+Export only one component per file as the default (eslint: [`react/no-multi-comp`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-multi-comp.md))
```
Comment (dcartereb) [inline react/README.md]: I think there could be good reasons to export related components that belong to a suite.  For example, `<Modal>` and `<ModalFooter>`.  You could define the footer contents as a prop, but being able to compose the footer in JSX is more expressive.  Also, the footer has no context outside of `<Modal>`, so it makes sense to me to have them in the same module.  I also foresee nested `<Grid>` components (`<Grid>`, `<GridCell>`, `<GridSqueezePlop>`, etc) being very useful and expressive, but inherently coupled.

Comment (BenAtEventbrite) [inline react/README.md]: Yeah, I totally agree that they are inherently coupled. I also agree that JSX is way more expressive than using props for these cases. However, as we've been thinking about it, we still want to try keeping them separate and see how that feels.

What's interesting though is that `Modal` & `ModalFooter` probably have different classifications. `Modal` may be a molecule while `ModalFooter` is an atom.

In the end, I think keeping them separate will really help force us into thinking more atomically, which I feel right now takes precedence over grouping them together in a file.

What do you think?

Comment (dcartereb) [inline react/README.md]: Seems fine to me.  So long as `<Modal>` and `<ModalFooter>` either appear together or have a link so it's easy to see the relationship.  By this I mean:  if I'm browsing molecules and I see `<Modal>`, I should see `<ModalFooter>` right there with it (even if it is considered an atom) so I know it exists and I don't try to recreate it.

Comment (BenAtEventbrite) [inline react/README.md]: Agreed. 👍

File: react/README.md [hunk 2]
```diff
@@ -18,3 +29,259 @@ For more info, see [Getting started with React](https://facebook.github.io/react
 [JSX](https://facebook.github.io/jsx/) is a JavaScript syntax extension that looks similar to XML. We use it with React because it is a concise and familiar syntax for defining tree structures with attributes.
 
 For more info, see [JSX in Depth](https://facebook.github.io/react/docs/jsx-in-depth.html).
+
+## General rules
+
+- Always use JSX syntax (don't use `React.createElement`)
```
Comment (vincent-eb) [inline react/README.md]: Maybe clarify a bit here what JSX syntax would look like / what exactly is the bad example an example of? I understand what you are saying but might be good for clarity.

Comment (BenAtEventbrite) [inline react/README.md]: good idea 👍

File: react/README.md [hunk 3]
```diff
@@ -18,3 +29,259 @@ For more info, see [Getting started with React](https://facebook.github.io/react
...
+    </Component>;
+}
+
+// bad (missing wrapping parentheses for single-line)
+let content = <div>Content</div>
+```
+
+## JSX alignment
+
+When a component has four props or less with no content, the tag can be on a single line (eslint: [`react/jsx-max-props-per-line`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-max-props-per-line.md)):
```
Comment (vincent-eb) [inline react/README.md]: I feel like 4 might even be too many, could cause inconsistency (hard to always keep that in mind / seems arbitrary). Maybe any more than two should be split to single line to increase similarity/readability between components? Just a thought

Comment (BenAtEventbrite) [inline react/README.md]: Yeah, I talked with @gagoar and we're going to go down to 3 for now. I think if we went with 2 then we might as well just never have single lines. 😄  I agree with the sentiment though. Just feel that something really short could end up being _less_ readable when forced to multiple lines.

File: react/README.md [hunk 4]
```diff
@@ -18,3 +29,276 @@ For more info, see [Getting started with React](https://facebook.github.io/react
...
+
+- When the component does more than just render `props`, such as attaching callback handlers, the stateless function can become unnecessarily complex with nested functions
+- [`propTypes`](#component-propTypes) are defined _within_ the ES6 class, but have to be defined as additional properties _outside_ of a stateless function
+- Using ES6 classes for the main/default component help differentiate it from [helper components](#)
+
+## Component organization
+
+Export only one component per file as the default (eslint: [`react/no-multi-comp`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-multi-comp.md))
+
+```js
```
Comment (jphilipps-eb) [inline react/README.md]: For these, you mean that we should only have one component per file, correct? Maybe you could add a commented out filename like `// MainComponent.js` on top of each to make that clearer?

**Comment by jphilipps-eb** [comment]
This looks great - I like seeing the `good` way of writing code as well as the multiple `bad` examples. This makes it super clear! 💯

**Comment by BenAtEventbrite** [comment]
Thanks @jphilipps-eb!

File: react/README.md (hunk 4 above)
Comment (BenAtEventbrite) [inline react/README.md]: Yep. Good idea 👍

**Pull request merged**